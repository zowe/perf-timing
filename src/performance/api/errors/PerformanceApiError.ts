/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

 /**
  * Base error for issues thrown by the Performance API.
  *
  * @external
  */
export class PerformanceApiError extends Error {
    /**
     * Construct the error.
     * @param message Message sent by the child.
     */
    constructor(message: string) {
        super(`Performance Error\n${message}`);
    }
}
