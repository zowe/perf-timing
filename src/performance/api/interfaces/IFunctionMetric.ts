export interface IFunctionMetric {
    name: string;
    calls: number;
    totalDuration: number; // Both in ms
    averageDuration: number;
}
