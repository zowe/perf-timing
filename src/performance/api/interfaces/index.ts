export * from "./IFunctionMetric";
export * from "./IFunctionTimer";
export * from "./IMeasurement";
export * from "./IMeasurementMetric";
export * from "./IMeasureTimer";
export * from "./IMetrics";
