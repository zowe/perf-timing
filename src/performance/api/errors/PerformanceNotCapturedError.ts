/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { PerformanceApiError } from "./PerformanceApiError";

/**
 * Thrown when metric output methods were called and performance was not enabled.
 *
 * @example
 * const perfApi = new PerformanceApi({enabled: false, packageUUID: "bogus"});
 *
 * // Calling these methods will cause this error to be thrown.
 * perfApi.getMetrics();
 * perfApi.getNodeTiming();
 * perfApi.getSysInfo();
 */
export class PerformanceNotCapturedError extends PerformanceApiError {
    constructor() {
        super("Unable to access performance metrics when performance monitoring is not enabled!");
    }
}
