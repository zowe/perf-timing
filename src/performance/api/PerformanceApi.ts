/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IPerformanceApi, IPerformanceApiManager } from "../manager/interfaces";
import {
    ICollectionObserver,
    IFunctionObserver,
    IMeasurementObserver,
    IMetric,
    IMetrics,
    INodeTiming,
    IPerformanceEntry,
    ISystemInformation
} from "./interfaces";

/**
 * A type that generically defines all collection observer maps that are defined
 * to the Performance API.
 *
 * @param T The type of observer stored in the map. The observer must be an
 *          {@link ICollectionObserver} type with entries adhering to the
 *          {@link IPerformanceEntry} interface.
 *
 * @internal
 */
type CollectionMap<T extends ICollectionObserver<IPerformanceEntry>> = Map<string, T>;

/**
 * The underlying api that provides hooks into
 * {@link https://nodejs.org/dist/latest-v10.x/docs/api/perf_hooks.html Node's Performance Timing APIs}.
 *
 * Use this class in place of relying on Node's experimental APIs.
 */
export class PerformanceApi implements IPerformanceApi {

    /**
     * Internal method for getting the errors that can be thrown. It is smart
     * and will only do the require once and cache the results for later calls.
     *
     * @internal
     */
    private static get _errors(): typeof PerformanceApi._errorImport {
        if (!PerformanceApi._errorImport) {
            PerformanceApi._errorImport = require("./errors");
        }

        return PerformanceApi._errorImport;
    }

    /**
     * Cache of loaded error objects. Only populated on first call to {@link PerformanceApi._errors}
     *
     * @internal
     */
    private static _errorImport: typeof import("./errors");

    /**
     * Aggregate all entries present in a {@link CollectionMap} into an array output.
     *
     * Each item present in the map will correspond to a single entry in the final array output. The
     * strings will now become the name property and each data point will be stored in the final
     * object. The statistical analysis present is generated from the IPerformanceEntry instances
     * present in the entries portion of that element.
     *
     *
     * @param map A {@link CollectionMap} of observers containing entries to process.
     * @param T The type representing the raw data stored in the observer. It is important that this
     *          data extends {@link IPerformanceEntry} for proper data processing.
     *
     * @returns An array of {@link IMetric} data points representing each key/value present in the
     *          CollectionMap.
     *
     * @internal
     */
    private static _aggregateData<T extends IPerformanceEntry>(map: CollectionMap<ICollectionObserver<T>>): Array<IMetric<T>> {
        // Get all observers present within the map.
        const observers = map.entries();
        const output: Array<IMetric<T>> = [];

        // Loop through each observer in the map
        for (const [name, observer] of observers) {
            let totalDuration = 0;

            // Calculate the total duration represented by all entries
            for (const entry of observer.entries) {
                totalDuration += entry.duration;
            }

            // Create the data point for the map using the key as the name
            output.push({
                name,
                calls: observer.entries.length,
                totalDuration,
                averageDuration: totalDuration / observer.entries.length,
                entries: observer.entries
            });
        }

        return output;
    }

    /**
     * Created for consistent reference to Node Performance Timing API.
     *
     * @see {@link PerformanceApi.watch}
     */
    public timerify = this.watch;

    /**
     * Created for consistent reference to Node Performance Timing API.
     *
     * @see {@link PerformanceApi.unwatch}
     */
    public untimerify = this.unwatch;

    /**
     * Internal map of all created function timers. The key represents the name
     * and the value is an {@link IFunctionObserver} instance.
     *
     * @internal
     */
    private _functionObservers: CollectionMap<IFunctionObserver> = new Map();

    /**
     * Internal map of all created measurement timers. The key represents the name
     * and the value is an {@link IMeasurementObserver} instance.
     *
     * @internal
     */
    private _measurementObservers: CollectionMap<IMeasurementObserver> = new Map();

    /**
     * This variable holds the import from the Node JS performance hooks
     * library.
     *
     * @internal
     */
    private readonly _perfHooks: typeof import("perf_hooks");

    /**
     * Construct the performance API. The constructor is not intended to be used
     * by any class except for the {@link PerformanceApiManager} of this class.
     *
     * @internal
     *
     * @param _manager The manager of this API class.
     */
    constructor(private readonly _manager: IPerformanceApiManager) {
        // Check if performance utilities should be enabled.
        if (this._manager.isEnabled) {
            // Delay the require so we don't waste resources when performance
            // isn't needed.
            this._perfHooks = require("perf_hooks");
        }
    }

    /**
     * Clears marks created using {@link mark}.
     *
     * This method does nothing if the Performance API is not enabled. It also takes advantage
     * of the namespace applied within mark; however, if name is not specified, all marks
     * will be cleared from all package.
     *
     * @param name The name of the mark to clear. If not specified, all marks are
     *             cleared.
     *
     * @see {@link https://nodejs.org/api/perf_hooks.html#perf_hooks_performance_clearmarks_name clearMarks()}
     */
    public clearMarks(name?: string) {
        if (this._manager.isEnabled) {
            // Add the namespace to the mark for uniqueness across packages
            this._perfHooks.performance.clearMarks(
                name ? this._addPackageNamespace(name) : undefined
            );
        }
    }

     /**
      * Aggregate metrics that have been captured for this instance of the API
      * and return them back in a data format.
      *
      * @returns An object representing metrics that have been prepared to be
      *          written to a file.
      *
      * @throws {@link PerformanceNotCapturedError} when the method is called and performance captures
      *                                             are not enabled.
      */
    public getMetrics(): IMetrics {
        if (this._manager.isEnabled) {
            // @TODO All metrics should be stopped before reporting
            return {
                functions: PerformanceApi._aggregateData(this._functionObservers),
                measurements: PerformanceApi._aggregateData(this._measurementObservers)
            };
        }
        else {
            throw new PerformanceApi._errors.PerformanceNotCapturedError();
        }
    }

    /**
     * Get the node timing information provided by the underlying node APIs.
     *
     * @returns The node timing information.
     *
     * @throws {@link PerformanceNotCapturedError} when the method is called and performance captures
     *                                             are not enabled.
     */
    public getNodeTiming(): INodeTiming {
        if (this._manager.isEnabled) {
            const timing = this._perfHooks.performance.nodeTiming;

            return {
                bootstrapComplete: timing.bootstrapComplete,
                duration: timing.duration,
                loopStart: timing.loopStart,
                loopExit: timing.loopExit,
                name: timing.name,
                nodeStart: timing.nodeStart,
                startTime: timing.startTime,
                v8Start: timing.v8Start
            };
        } else {
            throw new PerformanceApi._errors.PerformanceNotCapturedError();
        }
    }

    /**
     * Gathers and returns valuable system information to help understand the environment
     * where the data was captured.
     *
     * @returns System information formatted in an object ready to write to a file.
     *
     * @throws {@link PerformanceNotCapturedError} when the method is called and performance captures
     *                                             are not enabled.
     */
    public getSysInfo(): ISystemInformation {
        if (this._manager.isEnabled) {
            // Dynamically import os so that this process doesn't have to bother
            // if performance is not enabled.
            const os: typeof import("os") = require("os");

            const freeMem = os.freemem();
            const totalMem = os.totalmem();

            return {
                argv: process.argv,
                cpus: os.cpus(),
                loadavg: os.loadavg(), // On windows this will always be [0, 0, 0]
                memory: {
                    free: freeMem,
                    total: totalMem,
                    usage: totalMem - freeMem,
                    usagePercentage: ((totalMem - freeMem) / totalMem) * 100 // tslint:disable-line:no-magic-numbers
                },
                network: {
                    hostname: os.hostname(),
                    interfaces: os.networkInterfaces()
                },
                os: `${os.type()} ${os.arch()} ${os.release()}`,
                platform: os.platform(),
                shell: os.userInfo().shell,
                uptime: os.uptime()
            };
        } else {
            throw new PerformanceApi._errors.PerformanceNotCapturedError();
        }
    }

    public mark(name: string) {
        if (this._manager.isEnabled) {
            this._perfHooks.performance.mark(
                this._addPackageNamespace(name)
            );
        }
    }


    // @TODO document
    public measure(name: string, startMark: string, endMark: string) {
        if (this._manager.isEnabled) {
            let mapObject: IMeasurementObserver;

            if (this._measurementObservers.has(name)) {
                mapObject = this._measurementObservers.get(name);
            } else {
                // Create the map object when needed
                mapObject = {
                    observer: undefined,
                    isConnected: false,
                    entries: []
                };

                this._measurementObservers.set(name, mapObject);
            }

            // Remap the name to the unique package name. This will not be visible
            // in the output of the command, but it will allow 2 instances of this
            // package to not conflict accidentally in their naming.
            const namespaceName = this._addPackageNamespace(name);
            const namespaceStartMark = this._addPackageNamespace(startMark);
            const namespaceEndMark = this._addPackageNamespace(endMark);

            // If the map object is already connected, we don't need to do anything.
            // The metric will be picked up by the for loop in the observer.
            if (!mapObject.isConnected) {
                // Create the observer that will capture the metrics.
                mapObject.observer = new this._perfHooks.PerformanceObserver((list) => {
                    const entries = list.getEntriesByName(namespaceName);

                    // Loop through each entry and save the metrics
                    if (entries.length > 0) {
                        for (const entry of entries) {
                            mapObject.entries.push({
                                name: entry.name,
                                startTime: entry.startTime,
                                duration: entry.duration,
                                startMarkName: namespaceStartMark,
                                endMarkName: namespaceEndMark
                            });
                        }

                        mapObject.observer.disconnect();
                        mapObject.isConnected = false;
                    }
                });
                mapObject.observer.observe({ entryTypes: ["measure"], buffered: true });
                mapObject.isConnected = true;
            }

            this._perfHooks.performance.measure(namespaceName, namespaceStartMark, namespaceEndMark);
        }
    }

    public unwatch(fn: ((...args: any[]) => any), name?: string) {
        if (this._manager.isEnabled) {
            let timer = fn.name;

            // Extract the name of the function if necessary
            if (name != null) {
                timer = name;
            } else {
                // When timerifing(?) functions, node will prepend the function
                // name with timerified . This regex will strip out that name
                timer = /timerified (.*)/g.exec(timer)[1];
            }

            const timerRef = this._functionObservers.get(timer);

            if (timerRef !== undefined) {
                timerRef.observer.disconnect();
                timerRef.isConnected = false;
                return timerRef.originalFunction;
            } else {
                throw new PerformanceApi._errors.TimerDoesNotExistError(timer);
            }
        } else {
            return fn;
        }
    }

    public watch(fn: (...args: any[]) => any, name?: string) {
        if (this._manager.isEnabled) {
            // Check if we should use the function name or the passed name for tracking.
            if (name == null) {
                name = fn.name;
            }

            // @TODO allow for a name to be reconnected if the observer has been disconnected.
            // Throw an error if the timer already exists in the map.
            if (this._functionObservers.has(name)) {
                throw new PerformanceApi._errors.TimerNameConflictError(name);
            }

            // Create the observer and store it in the map.
            const observerObject: IFunctionObserver = {
                observer: undefined,
                isConnected: false,
                entries: [],
                originalFunction: fn
            };

            this._functionObservers.set(name, observerObject);

            // Create a function observer
            observerObject.observer = new this._perfHooks.PerformanceObserver((list) => {
                const entries = list.getEntriesByName(fn.name);

                for (const entry of entries) {
                    observerObject.entries.push(entry);
                }
            });

            observerObject.observer.observe({ entryTypes: ["function"], buffered: true });
            observerObject.isConnected = true;

            // Wrap the function in a timer
            return this._perfHooks.performance.timerify(fn);
        } else {
            // If performance is not enabled, we need to return the function
            // so original code doesn't get broken.
            return fn;
        }
    }

    private _addPackageNamespace(name: string): string {
        return `${this._manager.packageUUID}: ${name}`;
    }
}
