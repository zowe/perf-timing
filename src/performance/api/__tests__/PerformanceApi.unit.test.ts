/*!
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

jest.mock("perf_hooks");
jest.mock("os");

import { CollectionMap, PerformanceApi } from "../PerformanceApi";
import {
    ICollectionObserver,
    IFunctionObserver,
    IMeasurementObserver,
    IMetric,
    IPerformanceEntry,
    IPerformanceObserver
} from "../interfaces";
import { IPerformanceApiManager } from "../../manager/interfaces";

import * as perfHooks from "perf_hooks";
import * as os from "os";

import { randomHexString, randomInteger, randomNumber, getMockWrapper } from "../../../../__tests__/utilities";
import { PerformanceNotCapturedError } from "../errors";

/**
 * Strongly typed access to the private static methods of the PerformanceApi. For
 * documentation on these methods, see the corresponding item in the {@link PerformanceApi}.
 */
interface IPerformanceApiStaticPrivate {
    _errorImport: typeof import("../errors");
    readonly _errors: IPerformanceApiStaticPrivate["_errorImport"];
    _aggregateData<T extends IPerformanceEntry>(map: CollectionMap<ICollectionObserver<T>>): Array<IMetric<T>>;
}

type PerformanceApiType = { [K in keyof PerformanceApi]: PerformanceApi[K] };


/**
 * Strongly typed access to the private instance methods of the PerformanceApi. For
 * documentation on these methods, see the corresponding item in the {@link PerformanceApi}.
 */
interface IPerformanceApiPrivate extends PerformanceApiType {
    _functionObservers: CollectionMap<IFunctionObserver>;
    readonly _manager: IPerformanceApiManager;
    _measurementObservers: CollectionMap<IMeasurementObserver>;
    readonly _perfHooks: typeof import("perf_hooks");
    _addPackageNamespace(value: string): string;
}

/**
 * Gets a dummy api manager object for testing.
 *
 * @param packageUUID The uuid of the package manager
 * @param isEnabled Is the manager enabled
 *
 * @returns A manager object that can be passed to the api
 */
function getManager(packageUUID: string, isEnabled = true): IPerformanceApiManager {
    return {
        isEnabled,
        packageUUID
    };
}

/**
 * Generates a dummy performance observer for testing.
 *
 * @returns a dummy performance observer.
 */
function getDummyObserver(): IPerformanceObserver {
    return {
        disconnect: jest.fn(),
        observe: jest.fn()
    };
}

describe("PerformanceApi", () => {
    // This gives us an all access pass to the performance timing api.
    const _PerformanceApi: IPerformanceApiStaticPrivate = PerformanceApi as any;

    beforeEach(() => {
        // Clear out the value of the error import for each test.
        _PerformanceApi._errorImport = undefined;

        jest.clearAllMocks();
    });

    describe("private static functions", () => {
        it("should load errors on demand", () => {
            expect(_PerformanceApi._errorImport).toBeUndefined();

            const errors = require("../errors");

            expect(_PerformanceApi._errors).toBe(errors);
            expect(_PerformanceApi._errorImport).toBe(errors);

            // Tests the else path of the get function
            expect(_PerformanceApi._errors).toBe(errors);
        });

        describe("_aggregateData", () => {
            const generateCollectionMap = (mapData: Array<[string, IPerformanceEntry[]]>): CollectionMap<ICollectionObserver<IPerformanceEntry>> => {
                const map: ReturnType<typeof generateCollectionMap> = new Map();

                for (const data of mapData) {
                    map.set(data[0], {
                        entries: data[1],
                        isConnected: false,
                        observer: getDummyObserver()
                    });
                }

                return map;
            };

            it("should convert a collection map to the proper format. (1 map entry)", () => {
                const testMap = generateCollectionMap([
                    ["test element 1", [
                        {
                            duration: 5.325,
                            name: "test",
                            startTime: 12
                        }
                    ]]
                ]);

                // Hard coded test to see if we get the right data from a single
                // metric.
                expect(_PerformanceApi._aggregateData(testMap)).toEqual([
                    {
                        averageDuration: 5.325,
                        calls: 1,
                        entries: [{
                            duration: 5.325,
                            name: "test",
                            startTime: 12
                        }],
                        name: "test element 1",
                        totalDuration: 5.325
                    }
                ]);
            });

            it("should convert a randomized map to the proper output", () => {
                // tslint:disable:no-magic-numbers

                const numCollectionItems = randomInteger(5, 20);
                const expectedArray: Array<IMetric<IPerformanceEntry>> = [];
                const map = generateCollectionMap([]); // I'm too lazy to keep typing this

                // Create a random number of items in the map, between 5 and 20 will be created
                for(let itemCounter = 0; itemCounter < numCollectionItems; itemCounter++) {
                    // Create a unique key name for each item
                    const keyName = `${itemCounter}_${randomHexString(randomInteger(10, 20))}`;
                    let totalDuration = 0;

                    // Default to 0 entries
                    let min = 0;
                    let max = 0;

                    const entries: IPerformanceEntry[] = [];

                    if (itemCounter === 1) {
                        min = 1;
                        max = 1;
                    } else if (itemCounter > 1) {
                        min = 2;
                        max = 10;
                    }

                    // For item 0 there will be no entries so that might blow up.
                    // For item 1 there will be 1 entry
                    // For items 2+ there will be between 2 and 10 entries
                    const numEntries = randomInteger(min, max);
                    for (let entryCounter = 0; entryCounter < numEntries; entryCounter++) {
                        const insertObj = {
                            duration: randomNumber(0, 15),
                            startTime: randomNumber(0, 15),
                            name: randomHexString(randomInteger(10, 20))
                        };

                        // Keep track of the duration for this run
                        totalDuration += insertObj.duration;

                        entries.push(insertObj);
                    }

                    // Add the item to the map
                    map.set(keyName, {
                        entries,
                        isConnected: false,
                        observer: getDummyObserver()
                    });

                    // Now add the item to the expected array as well
                    expectedArray.push({
                        name: keyName,
                        calls: entries.length,
                        totalDuration,
                        averageDuration: totalDuration / entries.length,
                        entries
                    });
                }

                expect(_PerformanceApi._aggregateData(map)).toEqual(expectedArray);

                // tslint:enable:no-magic-numbers
            });

            it("should return an empty array when there is no data", () => {
                expect(_PerformanceApi._aggregateData(new Map())).toEqual([]);
            });
        });
    });

    describe("private instance functions/variables", () => {
        it("should properly initialize private variables (manager disabled)", () => {
            const manager = getManager("Malkin@71", false);

            const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            expect(api._functionObservers).toBeInstanceOf(Map);
            expect(api._functionObservers.size).toBe(0);

            expect(api._measurementObservers).toBeInstanceOf(Map);
            expect(api._measurementObservers.size).toBe(0);

            expect(api._perfHooks).toBeUndefined();
            expect(api._manager).toBe(manager);

        });
        it("should properly initialize private variables (manager enabled)", () => {
            const manager = getManager("Guentzel@59");

            const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            expect(api._functionObservers).toBeInstanceOf(Map);
            expect(api._functionObservers.size).toBe(0);

            expect(api._measurementObservers).toBeInstanceOf(Map);
            expect(api._measurementObservers.size).toBe(0);

            expect(api._perfHooks).toBe(require("perf_hooks"));
            expect(api._manager).toBe(manager);
        });

        it("should properly add the package namespace", () => {
            const manager = getManager("Letang@58", true);

            const _api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            for(const test of ["", "Value 1", "12345", "false"]) {
                expect(_api._addPackageNamespace(test)).toBe(`${manager.packageUUID}: ${test}`);
            }
        });
    });

    describe("gathering metrics", () => {
        let mocks = getMockWrapper({
            aggregateData: _PerformanceApi._aggregateData
        });

        beforeAll(() => {
            jest.spyOn(_PerformanceApi, "_aggregateData");
            mocks = getMockWrapper({
                aggregateData: _PerformanceApi._aggregateData
            });
        });
        beforeEach(() => {
            mocks.aggregateData.mockClear();
            mocks.aggregateData.mockImplementation(() => undefined);
        });
        afterAll(() => mocks.aggregateData.mockRestore());

        describe("getMetrics", () => {
            it("should return gathered metrics", () => pending());
            it("should throw PerformanceNotCapturedError", () => {
                const api = new PerformanceApi(getManager("NotEnabled", false));

                expect(() => {
                    api.getMetrics();
                }).toThrowError(PerformanceNotCapturedError);

                expect(_PerformanceApi._aggregateData).not.toHaveBeenCalled();
            });
        });

        describe("getNodeTiming", () => {
            Object.defineProperty(perfHooks.performance, "nodeTiming", {
                get() {
                    return {
                        bootstrapComplete: 0,
                        duration: 1,
                        loopStart: 2,
                        loopExit: 3,
                        name: "Some Name",
                        nodeStart: 4,
                        startTime: 5,
                        v8Start: 6
                    };
                },
                configurable: true
            });

            const spy = jest.spyOn(perfHooks.performance, "nodeTiming", "get");

            it("should return node timing information", () => {
                const api = new PerformanceApi(getManager("NodeTiming"));

                // Check that the resulting call is consistent
                expect(api.getNodeTiming()).toMatchSnapshot();
            });

            it("should throw PerformanceNotCapturedError", () => {
                const api = new PerformanceApi(getManager("NotEnabled", false));

                expect(() => {
                    api.getNodeTiming();
                }).toThrowError(PerformanceNotCapturedError);

                expect(spy).not.toHaveBeenCalled();
            });
        });

        describe("getSysInfo", () => {
            it("should return system information", () => {
                const osMock = getMockWrapper(os);

                const argv = process.argv;
                process.argv = ["a", "b", "c", "d"];

                osMock.cpus.mockReturnValue("CPU Information");
                osMock.loadavg.mockReturnValue("Load Avg Information");
                osMock.hostname.mockReturnValue("Hostname Information");
                osMock.networkInterfaces.mockReturnValue("Network Interfaces Information");
                osMock.type.mockReturnValue("OS Type Information");
                osMock.arch.mockReturnValue("OS Arch Information");
                osMock.release.mockReturnValue("OS Release Information");
                osMock.platform.mockReturnValue("OS Platform Information");
                osMock.userInfo.mockReturnValue({shell: "User Shell Information"});
                osMock.uptime.mockReturnValue("Uptime Information");

                osMock.freemem.mockReturnValue(1);
                osMock.totalmem.mockReturnValue(1);

                // First test that the object format is unchanged
                const api = new PerformanceApi(getManager("SysInfoTest"));
                expect(api.getSysInfo()).toMatchSnapshot();

                const testSet: Set<{free: number, total: number}> = new Set([
                    {free: 0, total: 0},
                    {free: 1, total: 0},
                    {free: 0, total: 1},
                    {free: 1, total: 1}
                ]);

                // Define some normal min/max ranges to be expected from os.freemem/os.totalmem
                const additionalTests = 50;
                const freeMin = 0;
                const freeMax = freeMin + additionalTests - 1;
                const totalMin = freeMax + 1;
                const totalMax = totalMin + additionalTests - 1;

                // Add more randomized tests
                for (let i = 0; i < additionalTests; i ++) {
                    testSet.add({
                        free: randomInteger(freeMin, freeMax),
                        total: randomInteger(totalMin, totalMax)
                    });
                }

                // Next do some testing with the memory calculations.
                for (const [{free, total}] of testSet.entries()) {
                    osMock.freemem.mockReturnValueOnce(free);
                    osMock.totalmem.mockReturnValueOnce(total);

                    const usage = total - free;
                    const usagePercentage = (usage / total) * 100; // tslint:disable-line:no-magic-numbers

                    expect(api.getSysInfo().memory).toEqual({
                        free,
                        total,
                        usage,
                        usagePercentage
                    });
                }

                process.argv = argv;
            });

            it("should throw PerformanceNotCapturedError", () => {
                const api = new PerformanceApi(getManager("NotEnabled", false));

                expect(() => {
                    api.getSysInfo();
                }).toThrowError(PerformanceNotCapturedError);

                // Just sanity check that one of the os functions wasn't called
                expect(os.freemem).not.toHaveBeenCalled();
            });
        });
    });

    describe("point to point measurement", () => {
        it("should do nothing when performance is not enabled", () => {
            const manager = getManager("PerformanceDisabled", false);
            const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            api.mark("Should not have marked");
            expect(perfHooks.performance.mark).not.toHaveBeenCalled();

            api.clearMarks("Should not have cleared marks");
            expect(perfHooks.performance.clearMarks).not.toHaveBeenCalled();

            api.measure("Should not have measured", "a", "b");
            expect(perfHooks.performance.measure).not.toHaveBeenCalled();
        });

        it("should clear all marks", () => {
            const manager = getManager("Jeff");
            const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            (api as any)._addPackageNamespace = jest.fn();

            api.clearMarks();

            expect(api._addPackageNamespace).not.toHaveBeenCalled();
            expect(perfHooks.performance.clearMarks).toHaveBeenCalledTimes(1);
            expect(perfHooks.performance.clearMarks).toHaveBeenCalledWith(undefined);
        });

        it("should clear a single mark", () => {
            const uuid = "Matt";
            const name = "Murray";

            const manager = getManager(uuid);
            const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            jest.spyOn(api, "_addPackageNamespace");

            api.clearMarks(name);

            expect(api._addPackageNamespace).toHaveBeenCalledWith(name);
            expect(perfHooks.performance.clearMarks).toHaveBeenCalledTimes(1);
            expect(perfHooks.performance.clearMarks).toHaveBeenCalledWith(`${uuid}: ${name}`);
        });

        it("should create a single mark", () => {
            const uuid = "Matt";
            const name = "Murray";

            const manager = getManager(uuid);
            const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

            jest.spyOn(api, "_addPackageNamespace");

            api.mark(name);

            expect(api._addPackageNamespace).toHaveBeenCalledWith(name);
            expect(perfHooks.performance.mark).toHaveBeenCalledTimes(1);
            expect(perfHooks.performance.mark).toHaveBeenCalledWith(`${uuid}: ${name}`);
        });

        describe("measure", () => {
            it("should measure between two marks", () => {
                // This should be a simple function to test that a measurement was called
                const manager = getManager("Mario");
                const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

                const name = "Measure Name";
                const startMark = "Start Mark";
                const endMark = "End Mark";

                jest.spyOn(api, "_addPackageNamespace");

                api.measure(name, startMark, endMark);

                expect(perfHooks.PerformanceObserver).toHaveBeenCalledTimes(1);
                expect(perfHooks.PerformanceObserver).toHaveBeenCalledWith(expect.any(Function));

                expect(api._addPackageNamespace).toHaveBeenCalledWith(name);
                expect(api._addPackageNamespace).toHaveBeenCalledWith(startMark);
                expect(api._addPackageNamespace).toHaveBeenCalledWith(endMark);

                const mapObj = api._measurementObservers.get(name);

                expect(mapObj.entries).toEqual([]);
                expect(mapObj.isConnected).toBe(true);
                expect(mapObj.observer).toBeInstanceOf(perfHooks.PerformanceObserver);
                expect(mapObj.observer.observe).toHaveBeenCalledTimes(1);
                expect(mapObj.observer.observe).toHaveBeenCalledWith({ entryTypes: ["measure"], buffered: true });

                expect(perfHooks.performance.measure).toHaveBeenCalledTimes(1);
                expect(perfHooks.performance.measure).toHaveBeenCalledWith(
                    api._addPackageNamespace(name),
                    api._addPackageNamespace(startMark),
                    api._addPackageNamespace(endMark)
                );
            });

            it("should not create another observer if one is connected", () => {
                // This should be a simple function to test that a measurement was called
                const manager = getManager("Mario");
                const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

                const name = "Measure Name";

                const startMark1 = "Start Mark 1";
                const endMark1 = "End Mark 1";

                const startMark2 = "Start Mark 2";
                const endMark2 = "End Mark 2";

                jest.spyOn(api, "_addPackageNamespace");
                jest.spyOn(api._measurementObservers, "set");

                api.measure(name, startMark1, endMark1);
                api.measure(name, startMark2, endMark2);

                expect(api._measurementObservers.set).toHaveBeenCalledTimes(1);
                expect(perfHooks.PerformanceObserver).toHaveBeenCalledTimes(1);

                expect(perfHooks.performance.measure).toHaveBeenNthCalledWith(
                    1,
                    api._addPackageNamespace(name),
                    api._addPackageNamespace(startMark1),
                    api._addPackageNamespace(endMark1)
                );

                expect(perfHooks.performance.measure).toHaveBeenNthCalledWith(
                    2,
                    api._addPackageNamespace(name),
                    api._addPackageNamespace(startMark2),
                    api._addPackageNamespace(endMark2)
                );
            });

            it("should collect metrics from the observer", () => {
                // tslint:disable:no-magic-numbers
                // This should be a simple function to test that a measurement was called
                const manager = getManager("Mario");
                const api: IPerformanceApiPrivate = new PerformanceApi(manager) as any;

                const name = "Measure Name";
                const startMark = "Start Mark";
                const endMark = "End Mark";

                api.measure(name, startMark, endMark);

                const mapObj = api._measurementObservers.get(name);

                const observerFunction = (perfHooks.PerformanceObserver as jest.Mock<any>).mock.calls[0][0];
                const listObj = {
                    getEntriesByName: jest.fn()
                };

                listObj.getEntriesByName.mockReturnValue([]);

                observerFunction(listObj);

                expect(mapObj.isConnected).toBe(true);
                expect(mapObj.observer.disconnect).not.toHaveBeenCalled();

                expect(listObj.getEntriesByName).toHaveBeenCalledTimes(1);
                expect(listObj.getEntriesByName).toHaveBeenCalledWith(api._addPackageNamespace(name));

                // Perform a second measure and give it data this time
                const entries = [
                    {
                        name,
                        startTime: 123,
                        duration: 456
                    },
                    {
                        name,
                        startTime: 789,
                        duration: 123
                    }
                ];

                listObj.getEntriesByName.mockReturnValueOnce(entries);

                observerFunction(listObj);

                expect(mapObj.isConnected).toBe(false);
                expect(mapObj.observer.disconnect).toHaveBeenCalled();

                // Check that the data object output remains unchanged.
                expect(mapObj.entries).toMatchSnapshot("measure data collection is unchanged");

                // tslint:enable:no-magic-numbers
            });
        });
    });

    describe("watching a function", () => {
        it("should map timerify to watch", () => {
            const manager = getManager("ABCDE", false);

            const api = new PerformanceApi(manager);

            expect(api.timerify).toBe(api.watch);
        });

        it("should map untimerify to unwatch", () => {
            const manager = getManager("ABCDE", false);

            const api = new PerformanceApi(manager);

            expect(api.untimerify).toBe(api.unwatch);
        });

        describe("watching a function", () => {
            it("should watch a function", () => pending());
            it("should return the proper function", () => pending()); // Handle enabled vs disabled here
            it("should gather metrics on the watched function", () => pending());
            it("should throw a TimerNameConflictError", () => pending());
            it("should allow the timer to be saved under a different name", () => pending()); // Note this might cause problems with 2
                                                                                               // functions of the same name so we get to test it
        });

        describe("unwatching a function", () => {
            it("should return the input function", () => pending());
            it("should return the original function implementation", () => pending());
            it("should disconnect the observer", () => pending()); // Test that calling the function doesn't trigger the observer
            it("should accept the name of the timer that is unwatched", () => pending());
        });
    });
});
