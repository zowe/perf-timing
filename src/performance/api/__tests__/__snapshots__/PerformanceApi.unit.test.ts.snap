// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PerformanceApi point to point measurement measure should collect metrics from the observer: measure data collection is unchanged 1`] = `
Array [
  Object {
    "duration": 456,
    "endMarkName": "Mario: End Mark",
    "name": "Measure Name",
    "startMarkName": "Mario: Start Mark",
    "startTime": 123,
  },
  Object {
    "duration": 123,
    "endMarkName": "Mario: End Mark",
    "name": "Measure Name",
    "startMarkName": "Mario: Start Mark",
    "startTime": 789,
  },
]
`;
